scripts/shared.sh

```diff
@@ -4,11 +4,106 @@
 set -a
 set -eu -o pipefail
 
+function _extract_from_file() {
+    local key="$1"
+    local config_file="$2"
+    local value
+    if [[ -f $config_file ]]; then
+        value="$(eval "jq -r .${key} $config_file")"
+        if [[ "$value" != "null" ]]; then
+            echo "$value"
+            return 0
+        fi
+    fi
+}
+
 function _extract_cluster_config_value() {
     # Try extracting config from cluster config.json, default config.json, then aladdin config.json
+    local key="$1"
     local value
-    value="$1"
-    jq -nr --arg value "$value" 'first(inputs | (if .[$value] == null then empty else .[$value] end))' \
-        "$ALADDIN_CONFIG_DIR/$CLUSTER_CODE/config.json" "$ALADDIN_CONFIG_DIR/default/config.json" \
-        "$ALADDIN_CONFIG_DIR/config.json"
+    value=$(_extract_from_file "$key" "$ALADDIN_CONFIG_DIR/$CLUSTER_CODE/config.json")
+    if test -n "$value"; then
+        echo "$value"
+        return 0
+    fi
+    value=$(_extract_from_file "$key" "$ALADDIN_CONFIG_DIR/default/config.json")
+    if test -n "$value"; then
+        echo "$value"
+        return 0
+    fi
+    value=$(_extract_from_file "$key" "$ALADDIN_CONFIG_DIR/config.json")
+    if test -n "$value"; then
+        echo "$value"
+        return 0
+    fi
+}
+
+function time_plus_offset() {
+    local offset="$1"
+    local current_time="$(date +'%s')"
+    echo "$((${current_time}+offset))"
+}
+
+function which_exists(){
+    for cmd in "$@" ; do
+        if which $cmd &>/dev/null ; then
+            echo "$cmd"
+            return 0
+        fi
+    done
+    return 1
+}
+
+function clear_cache_file(){
+    rm -f "$HOME/.aladdin/infra/cache.json"
+}
+
+function set_cache() {
+    local key="$(make_hash $1)"
+    local expiration="${2:-}"
+    local data="${3:-}"
+    _get_or_set_cache $key $expiration $data
+}
```

Comment:
the arguments seem kind of backwards. generally caching APIs have the TTL as optional. i don't think it would be overly burdensome to require the second argument if the TTL is needed. if we really wanted to mark something as value-less for the sake of checking an expired time (like the aladdin image refresh, i might suggest a separate set of functions like `set_ttl` and `check_ttl` that just take a name and a TTL (when setting)

Reply:
fixed 👍